cpp/marc_tools/single_marc_convert_keywords.cc

```diff
@@ -0,0 +1,167 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <fstream>
+#include <iostream>
+#include <set>
+#include <stdexcept>
+#include <utility>
+#include <climits>
+#include <cstdio>
+#include <cstdlib>
+#include "FileUtil.h"
+#include "MARC.h"
+#include "StringUtil.h"
+#include "TextUtil.h"
+#include "util.h"
+
+
+namespace {
+
+
+[[noreturn]] void Usage() {
+    ::Usage("gnd_input keyword_input keyword_matches_output keyword_without_matches_output"
+               "       Searches for keyword matches in the \"gnd_input\" file."
+               "       Returns a \"keyword_matches_output\" file with matching keywords and their PPN,"
+               "       as well as \"keywords_without_matches\" file containing keywords where no matches were found.\n");
+}
+
+
+void extractSubfieldsForTag(MARC::Record const &record, const std::string &subfield_tag, const std::string &subfield_codes, std::vector<std::string> *subfields) {
+    if (not record.hasTag(subfield_tag))
+        return;
+    for (const char &subfield_code : subfield_codes) {
+        std::string subfield_value(record.getFirstSubfieldValue(subfield_tag, subfield_code));
+        if (not subfield_value.empty() ) {
+            if (subfield_code == 'x')
+                subfield_value = '(' + subfield_value + ')';
+            subfields->emplace_back(subfield_value);
+        }
+        else if (subfield_code == 'a')
+            LOG_WARNING("Entry has no Subfield 'a' for PPN " + record.getControlNumber());
+    }
+}
+
+
+void ReadInGndKeywords(MARC::Reader * const marc_reader, std::unordered_map<std::string, std::string> * const gnd_keywords)
+{
+    unsigned record_count(0);
+
+    while (MARC::Record record = marc_reader->read()) {
+        ++record_count;
+
+        std::vector<std::string> subfields;
+        extractSubfieldsForTag(record, "150", "agx", &subfields);
+        //entry for every single 'a' subfield
+        if (subfields.size() == 1)
+            gnd_keywords->emplace(std::make_pair(subfields[0], record.getControlNumber()));
+        //entry for all subfields combined
+        if (subfields.size() > 1) {
+            std::string key(StringUtil::Join(subfields, " "));
+            gnd_keywords->emplace(std::make_pair(key , record.getControlNumber()));
+        }
+    }
+    
+    LOG_INFO("Processed " + std::to_string(record_count) + " MARC record(s).");
+}
+
+
+void FindEquivalentKeywords(std::unordered_map<std::string, std::string> const *keywords_to_gnd, std::unordered_set<std::string> const &keywords_to_compare, const std::string matches_output_file, const std::string no_matches_output_file) {
```

Comment:
This seems fishy!  It seems that you should pass a reference and not a pointer here.